z3/src/ast.rs

```diff
@@ -196,6 +196,25 @@ pub trait Ast<'ctx>: Sized + fmt::Debug + Into<Dynamic<'ctx>> {
         })
     }
 
+
+    /// Compare this `Ast` with another `Ast`, and get a Result.  Errors if the sort does not
+    /// match for the two values.
+    fn _safe_eq(&self, other: &Self) -> Result<Bool<'ctx>, std::string::String> {
+        let ctx = self.get_ctx().z3_ctx;
+        let other_ctx = other.get_ctx().z3_ctx;
+        let left_sort = unsafe {
+            Z3_get_sort_kind(ctx, Z3_get_sort(ctx, self.get_z3_ast()))
+        };
+        let right_sort = unsafe {
+            Z3_get_sort_kind(other_ctx, Z3_get_sort(other_ctx, other.get_z3_ast()))
+        };
+        if (left_sort == right_sort) {
+            Ok(self._eq(other))
+        } else {
+            Err(format!("Can not compare nodes, Sort does not match.  Nodes contain types {:?} and {:?}", left_sort, right_sort))
```

Comment:
I agree that it's implied that the two don't have the same sort but I don't think the Err type should be a string. Maybe better suited as a struct whose display trait impl is the error message. Like

```rust
struct SortDiffers<'ctx> {
  left: Sort<'ctx>, right: Sort<'ctx>
}

impl<'ctx> Display for Sort<'ctx> { ... }
```